### PR TITLE
ENH: make "closed" part of IntervalDtype

### LINKED ISSUE #19371: API: Add closed to IntervalDtype
This was brought up by @shoyer in a different issue, and it makes sense to me

Adding `closed` to `IntervalDtype` makes sense to me, as in my mind `closed` is inherently part of the dtype.  Since differing `closed` makes two `IntervalIndex` incompatible, I'd expect their dtypes to not be equal, but they currently are, which seems a little strange:
```python
In [2]: ii1 = pd.interval_range(0, 3, closed='left')

In [3]: ii2 = pd.interval_range(0, 3, closed='right')

In [4]: ii1.dtype == ii2.dtype
Out[4]: True
```

There's also a larger discussion as to if adding `closed` to `IntervalDtype` would allow us to remove `closed` as a parameter to the `IntervalIndex` constructor.  My preference would be to keep the `closed` parameter for user convenience, similar to how `CategoricalIndex` accepts `categories` and `ordered` parameters despite also accepting `CategoricalDtype` (though maybe this is just for legacy reasons).  Willing to be convinced otherwise though.

xref https://github.com/pandas-dev/pandas/issues/19263#issuecomment-359986104 (original comment)
xref #19370 (might be rendered moot)

cc @shoyer

### PR DESCRIPTION
**Pull request by jbrockmendel**
- [x] closes #19371
- [x] tests added / passed
- [x] passes `black pandas`
- [x] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [x] whatsnew entry

Most of the work here is in accomodating backwards-compat.

### CODE REVIEW

File: pandas/core/arrays/interval.py [hunk 1]
```diff
@@ -268,6 +275,9 @@ def _simple_new(
             # If these share data, then setitem could corrupt our IA
             right = right.copy()
 
+        dtype = IntervalDtype(left.dtype, closed=closed)
+        result._dtype = dtype
+
```
Comment (jreback) [inline pandas/core/arrays/interval.py:280]: should _closed now be unnecessary? as it can directly ref the dtype

Comment (jbrockmendel) [reply pandas/core/arrays/interval.py:280]: you're right, will update

**Comment by jbrockmendel** [comment]
added whatsnew for deprecation

File: pandas/core/arrays/interval.py [hunk 2]
```diff
@@ -146,7 +146,7 @@
     >>> pd.arrays.IntervalArray([pd.Interval(0, 1), pd.Interval(1, 5)])
     <IntervalArray>
     [(0, 1], (1, 5]]
-    Length: 2, closed: right, dtype: interval[int64]
+    Length: 2, closed: right, dtype: interval[int64, right]
```
Comment (jreback) [inline pandas/core/arrays/interval.py:149]: should prob change the output formatting as well (e.g. closed: right) is now superfluous

File: pandas/core/dtypes/dtypes.py [hunk 3]
```diff
@@ -1002,37 +1003,69 @@ class IntervalDtype(PandasExtensionDtype):
...
 
-    def __new__(cls, subtype=None):
+    def __new__(cls, subtype=None, closed: Optional[str_type] = None):
         from pandas.core.dtypes.common import is_string_dtype, pandas_dtype
 
+        if closed is not None and closed not in {"right", "left", "both", "neither"}:
+            raise ValueError("closed must be one of 'right', 'left', 'both', 'neither'")
+
         if isinstance(subtype, IntervalDtype):
+            if closed is not None and closed != subtype.closed:
```
Comment (jreback) [inline pandas/core/dtypes/dtypes.py:1031]: test for this?

File: pandas/core/dtypes/dtypes.py [hunk 4]
```diff
@@ -1047,14 +1080,32 @@ def __new__(cls, subtype=None):
             )
             raise TypeError(msg)
 
+        if closed is None and subtype is not None:
```
Comment (jreback) [inline pandas/core/dtypes/dtypes.py:1083]: if we have this one do we need to warn on L1058? (e.g. which one is getting hit)

File: pandas/core/dtypes/dtypes.py [hunk 5]
```diff
@@ -1128,6 +1181,15 @@ def __setstate__(self, state):
         # PandasExtensionDtype superclass and uses the public properties to
         # pickle -> need to set the settable private ones here (see GH26067)
         self._subtype = state["subtype"]
+        # backward-compat older pickles won't have "closed" key
+        self._closed = state.pop("closed", None)
+        if self._closed is None:
+            warnings.warn(
+                "Unpickled legacy IntervalDtype does not specify 'closed' "
```
Comment (jreback) [inline pandas/core/dtypes/dtypes.py:1188]: test for this?

**Review by jreback — CHANGES_REQUESTED** [review]
looks good a few questions. biggest issue is the II repr is now duplicative. should fix here, though would increase the diff a lot, followon ok.

**Comment by jreback** [comment]
cc @jschendel if you can have a look.

File: pandas/core/arrays/interval.py (hunk 2 above)
Comment (jbrockmendel) [reply pandas/core/arrays/interval.py:149]: agreed

File: pandas/core/dtypes/dtypes.py (hunk 5 above)
Comment (jbrockmendel) [reply pandas/core/dtypes/dtypes.py:1194]: Yes, we have tests for all of the newly-added warning/raising cases in the IntervalDtype constructor.  (just added a missing consistency check https://github.com/pandas-dev/pandas/pull/38394/files#diff-f99ef42afad339d00e36197a60ccc76d74c6a94c30e05aef69d18e0ec4b10d4eR1055)

Comment (jbrockmendel) [reply pandas/core/dtypes/dtypes.py:1194]: woops, dont have a test for this one.  will update.

Comment (jbrockmendel) [reply pandas/core/dtypes/dtypes.py:1194]: huh, this is definitely reached, but tm.assert_produces_warning isnt seeing anything

**Comment by jreback** [comment]
> Performing static analysis using mypy
pandas/core/dtypes/dtypes.py:1250: error: Item "ExtensionDtype" of "Union[Any, ExtensionDtype]" has no attribute "closed"  [union-attr]
pandas/core/dtypes/dtypes.py:1251: error: Item "ExtensionDtype" of "Union[Any, ExtensionDtype]" has no attribute "closed"  [union-attr]

**Comment by jbrockmendel** [comment]
There's an annoying new failure mode:

```
dtype = pd.IntervalDtype("interval[int64]")   # <-- now warns and sets closed="right" as default

idx = pd.IntervalIndex([], dtype=dtype, closed="left")   # <-- raises ValueError bc of mismatch
```

File: pandas/core/dtypes/dtypes.py (hunk 4 above)
Comment (jreback) [reply pandas/core/dtypes/dtypes.py:1089]: can you respond to this

Comment (jbrockmendel) [reply pandas/core/dtypes/dtypes.py:1089]: they are both hit separately

**Review by jreback — APPROVED** [review]

**Comment by jreback** [comment]
lgtm. @jorisvandenbossche @jschendel if you can take a look.

**Comment by jorisvandenbossche** [comment]
Why is it needed to deprecate the default value for `closed`?

**Comment by jbrockmendel** [comment]
> Why is it needed to deprecate the default value for closed?

ATM there is (presumably) code in the wild calling the IntervalDtype constructor without specifying any `closed`.  Eventually failure to specify closed will raise.  Needs a deprecation cycle.

**Comment by jorisvandenbossche** [comment]
Yes, but so the question is: why does it need to eventually raise? 
Other constructors like `Interval` and `interval_range` also have a default. 

Another option could also be to have it "not initialized" as default, to enable inferring it from the passed Interval objects

**Comment by jbrockmendel** [comment]
> Another option could also be to have it "not initialized" as default, to enable inferring it from the passed Interval objects

id be OK with this

**Comment by jreback** [comment]
> > Another option could also be to have it "not initialized" as default, to enable inferring it from the passed Interval objects
> 
> id be OK with this

then the default should be 'infer' (which we can do if we have an unitialized value)

**Comment by jbrockmendel** [comment]
ok ill give this a shot and update

**Comment by jreback** [comment]
ping when ready here

**Comment by jreback** [comment]
this was ok, can you merge master

**Comment by jbrockmendel** [comment]
Most recent discussion is that I need to do a partial re-write, which I haven't gotten around to yet.

**Comment by jreback** [comment]
> Most recent discussion is that I need to do a partial re-write, which I haven't gotten around to yet.

ok!

**Comment by jbrockmendel** [comment]
updated as discussed.  agreed this is much nicer

**Comment by jreback** [comment]
I must be missing it, where exactly is the FutureWarning in the Dtype constructor triggered?

**Comment by jreback** [comment]
this no longer deprecates?

**Comment by jbrockmendel** [comment]
> this no longer deprecates?

correct.  passing "closed" to IntervalDtype is optional now

**Review by jreback — CHANGES_REQUESTED** [review]
lgtm. I would add a whatsnew note in other enhancements or create an interval bug fix section and say that the repr for Interval is changing.

**Comment by jbrockmendel** [comment]
> say that the repr for Interval is changing.

let's save that for the follow-up that changes that is aimed specifically at the repr

**Review by jreback — APPROVED** [review]

**Pull request merged**